src/Razor/src/Microsoft.CodeAnalysis.Razor.Workspaces/ProjectSystem/FallbackProjectManager.cs

```diff
@@ -101,15 +112,27 @@ private void AddFallbackProject(ProjectId projectId, string filePath)
     private void AddFallbackDocument(ProjectKey projectKey, string filePath, string projectFilePath)
     {
         var hostDocument = CreateHostDocument(filePath, projectFilePath);
+        if (hostDocument is null)
+        {
+            return;
+        }
+
         var textLoader = new FileTextLoader(filePath, defaultEncoding: null);
         _projectSnapshotManagerAccessor.Instance.DocumentAdded(projectKey, hostDocument, textLoader);
     }
 
-    private static HostDocument CreateHostDocument(string filePath, string projectFilePath)
+    private static HostDocument? CreateHostDocument(string filePath, string projectFilePath)
     {
-        var targetPath = filePath.StartsWith(projectFilePath, FilePathComparison.Instance)
-            ? filePath[projectFilePath.Length..]
-            : filePath;
+        // The compiler only supports paths that are relative to the project root, so filter our files
+        // that don't match
+        var projectPath = FilePathNormalizer.GetNormalizedDirectoryName(projectFilePath);
+        var normalizedFilePath = FilePathNormalizer.Normalize(filePath);
+        if (!normalizedFilePath.StartsWith(projectPath, FilePathComparison.Instance))
```

Comment:
It would be useful at some point later to create a `FilePathNormalizer.AreDirectoriesEquivalent` helper (similar to `FilePathsEquivalent`) to avoid having to create the strings at the end. It should be possible to do that without any allocations at all.